### PR TITLE
Implement LaTeX math support (WIP; Please do not merge this pull request yet)

### PR DESCRIPTION
**Pull request by dakusui**
This is a pull request that offers an implementation of Issue-#39 (LaTeX math mode support) 
**NOTE:** Please do not look into this pull request anymore. Refer to [this one](https://github.com/dakusui/ditaa/pull/3) for review.
**NOTE:** This pull request is opened for review and discussions. Another pull request will be created when it becomes necessary.

# README
Following is a documentation about this feature. This will appear in README.md after discussions are concluded.
### Command line
(snip)
```
 -e,--encoding <ENCODING>       The encoding of the input file.
 -h,--html                      In this case the input is an HTML file.
                                The contents of the <pre
                                class="textdiagram"> tags are rendered as
                                diagrams and saved in the images directory
                                and a new HTML file is produced with the
                                appropriate <img> tags.
    --help                      Prints usage help.
-L,--latex-math                 Enable LaTeX math mode.
```
(snip)
### Syntax
(snip)
#### LaTeX mode.
If you place LaTeX formulae inside 2 ```$```s, it will be rendered using [```jlatexmath```](https://github.com/opencollab/jlatexmath). That is, if you have a following input files.
```ditaa

$Box_1$                    $Box^2$
+---------------------+    +------+   /---------\
|$\sum_{i=0}^{n}x^i$  |    |$cBLU$|   |         |
|                     +--->|cRED  +-=-+cGRE$C_k$|
|{io}                 |    |cXYZ  |   |{o}      |
+----------+----------+    +---+--+   \---------/
           |                   |
           |                   :
           |                   V
           |           +-------------------+
           +---------->*$A_i$ hello $B^i$  |
                       |              +----+
                       |              |c8FA|
                       +--------------+----+

$|Set| = o-*-Freunde-*-nicht=*=diese-=-*- * töne$

o Quick brown fox jumps over
* a lazy dog.

$Q_u^i$, $C_k$, $B_r^{own}$, $F_{ox}$ jumps

over a lazy $d\cdot\frac{o}{g}$.


$\forall x \in X, \quad \exists y \leq \epsilon$


$\sin A \cos B =$

    $ \frac{1}{2}\left[ \sin(A-B)+\sin(A+B) \right]$


$\frac{d}{dx}\left( \int_{0}^{x} f(u)\,du\right)=f(x).$


 $v \sim \mathcal{N} (m,\sigma^2)$
```

This will be rendered as follows.

art-latexmath-1

##### Limitations
This feature is only available when you are generating ```.png``` files.

# Discussions
* This enhancement introduces a new semantics to a certain character in a text and it means potentially hurts a compatibility. I think I can implement a new option in command line to enable/disable this feature if you think it is important. Please let me know your thought.
* I think ```.svg``` support should be done in a separate pull-request since it would be a bit big change.
# Tasks
- [ ] Implement tests
- [x] Fix a bug where ```*``` is rendered as a bullet point (or a connector?) even if it is placed inside ```$``` and ```$```.
- [ ] Remove ```pom.xml```
  - [ ] Backport chages made in ```pom.xml``` to ```project.clj```
- [ ] Provide a documentation.
- [x] Implement an option to enable this feature. This must be disabled by default.

### CODE REVIEW

File: src/java/org/stathissideris/ascii2image/graphics/BitmapRenderer.java [hunk 1]
```diff
@@ -19,15 +19,14 @@
...
-import java.awt.RenderingHints;
-import java.awt.Stroke;
+import org.stathissideris.ascii2image.core.ConversionOptions;
+import org.stathissideris.ascii2image.core.RenderingOptions;
+import org.stathissideris.ascii2image.core.Shape3DOrderingComparator;
+import org.stathissideris.ascii2image.core.ShapeAreaComparator;
+import org.stathissideris.ascii2image.text.TextGrid;
+
+import javax.imageio.ImageIO;
+import java.awt.*;
```
Comment (stathissideris) [inline src/java/org/stathissideris/ascii2image/graphics/BitmapRenderer.java]: Please keep the explicit imports from before, it will help with the transition to Clojure.

File: src/java/org/stathissideris/ascii2image/text/TextGrid.java [hunk 2]
```diff
@@ -1,1786 +1,2094 @@
...
+          return true;
+      }
+    }
+    return false;
+  }
+
+
+  public CellSet getAllNonBlank() {
+    CellSet set = new CellSet();
+    int w
```
Comment (stathissideris) [inline src/java/org/stathissideris/ascii2image/text/TextGrid.java]: Fixing whitespace is great, but put such changes in a different pull request as they obscure the substance of your contribution and make the diff harder to read

**Comment by stathissideris** [comment]
1. We can continue work with this PR, new commits to your branch will be included, no need for a separate PR
2. I would like to keep backwards compatibility for the syntax, so if this feature introduces new semantics, it would need to be disabled by default and we should have a command-line option to enable it. Having it enabled by default would not work because some people use ditaa in an automated context and it may be hard for them to go and fix all invocations (and also it will create workload for them because they would have to check that everything renders the same way as before).
3. We would also need some documentation to demonstrate this great feature too!

File: src/java/org/stathissideris/ascii2image/graphics/BitmapRenderer.java (hunk 1 above)
Comment (dakusui) [reply src/java/org/stathissideris/ascii2image/graphics/BitmapRenderer.java]: Ok, apparently my IDE's setting is too much aggressive at formatting imports, let me fix it.

File: src/java/org/stathissideris/ascii2image/text/TextGrid.java (hunk 2 above)
Comment (dakusui) [reply src/java/org/stathissideris/ascii2image/text/TextGrid.java]: Got it, too. I'm fixing it soon. For the time being, please resort to "Hide white space changes", which is found under "Diff settings" button on a pull-request screen.

**Comment by dakusui** [comment]
1. Thank you for that.
2. Understood. I'll try to implement such an option which defaults to keep compatible behaviour.
3. I've added an action item for that and the section "LaTeX mode" in the description will be the documentation and it should be integrated in "README.md" of ditaa in this pull request. Please feel free to edit it for improvement and give me feedbacks.

**Comment by dakusui** [comment]
@stathissideris , I tried hard to fix the code formatting issue, but I found it's quite a bit painful. Maybe we should reformat the entire project first and resume working on this feature if it's hard to review even if we use "Ignore white space changes" feature of GitHub.
Can you please provide a code formatting style for some IDE such as Eclipse or IntelliJ? Or any style guideline if you have?

**Comment by stathissideris** [comment]
I have stopped working with Java for many years, so I don't have a preferred style. Are you using the Intellij defaults?

**Comment by stathissideris** [comment]
2 characters and spaces sounds like a reasonable default to me. Could you please send me a *new* PR on the current master that applies that to the whole project? Then you can git rebase this PR on the new master and continue work. Thanks!

**Comment by dakusui** [comment]
Understood, I have opened this pull request #43 for it.

**Comment by dakusui** [comment]
I've created another branch that contains this branch and pull request which is directed to a code formatted branch.
The diff is available [here](https://github.com/dakusui/ditaa/pull/3/files?utf8=%E2%9C%93&diff=split&w=1). I think this is reviewable and we should create a new pull request to be merged eventually.

**Comment by dakusui** [comment]
Please do not look into this pull request anymore. Refer to [this one](https://github.com/dakusui/ditaa/pull/3) for review.

**Comment by dakusui** [comment]
@stathissideris I have invited you to my forked repository as a collaborator for the review's sake. Please find an email from github or let me know to send it to you.
